### PR TITLE
Allow retroactive edits of existing CarePlan

### PR DESCRIPTION
**Pull request by hannahburkhardt**
- Allow retroactive edits of existing CarePlan https://www.pivotaltracker.com/story/show/184551787
- Fix incorrect birthday message date/time https://www.pivotaltracker.com/story/show/182907996

### Testing instructions for CarePlan editing
1. Open Enrollment client for a patient that doesn't have a CarePlan yet. You should be taken to the new Enrollment client like normal.
- test patient note still works
- test that adding and removing message list entries still works
- the CarePlan and CommunicationRequest resources that are created (when hitting done) match what you entered. No extraneous resources are created (no “revoked” CommunicationRequests are generated)
- cannot save the plan with empty messages
2. Open Enrollment client for a patient that does have a CarePlan already. When prompted, click “revoke and create new”.
- the old CarePlan is marked “revoked”
- all (not yet executed) CommunicationRequest resources from old CarePlan are marked “revoked”
- CommunicationRequests that were already executed do not change
- Go to 1.
3. Open Enrollment client for a patient that does have a CarePlan already. When prompted, click “edit”.
- patient note shows what the existing CarePlan had. If you change it and save, the existing CarePlan reflects this change (description attribute)
- The message list from the previously created CarePlan is shown, with correct dates/times and content.
- If removing any messages, once you save:
  - the corresponding CommunicationRequest resources are marked “revoked”
  - the corresponding entry in CarePlan.activity has been removed
- Messages that have already been sent (CommunicationRequest resources with status “completed”) are marked as such visually, and cannot be removed
- If changing any message date or text, the corresponding CommunicationRequest resources reflect the changes

### CODE REVIEW

File: src/components/EnrollmentApp.tsx [hunk 1]
```diff
@@ -1,7 +1,223 @@
...
+                    (value) => console.log(`CommunicationRequest/${cr.id} successfully revoked:`, value),
+                    //onrejected
+                    (rejectionReason) => console.log(`CommunicationRequest/${cr.id} revocation rejected:`, rejectionReason)
+                ).catch(
+                    (errorReason) => console.log(`CommunicationRequest/${cr.id} revocation caused error:`, errorReason)
+                );
+            });
+
+        }).then(() => {
+            existingCarePlan.status = "revoke";
```
Comment (achen2401) [inline src/components/EnrollmentApp.tsx:104]: @hannahburkhardt  I tested this in Smart Health IT launcher.  When I tried clicking on `revoke and create new` to revoke existing CommunicationRequests and CarePlan, it **worked** for **CommunicationRequests** - they were marked as revoked in the status field as expected.  However _it didn't_ for **CarePlan**.  This is the error message I got for CarePlan:  `Error was: Invalid attribute value \"revoke\": Unknown CarePlanStatus code 'revoke'`  I see allowed values for status is slightly different for [CarePlan](https://fhir-ru.github.io/careplan.html) vs [CommunicationRequest](https://fhir-ru.github.io/communicationrequest.html).

File: src/components/ScheduleSetup.tsx [hunk 2]
```diff
@@ -231,75 +261,62 @@ export default class ScheduleSetup extends React.Component<ScheduleSetupProps, S
...
-        );
+        </ListItem>
     }
 
     const removeMessage = (index: number) => {
-        let messages = props.messages;
-        messages.splice(index, 1);
-        props.onMessagesChanged(messages);
+        let messagePlan = props.messagePlan;
+        messagePlan.removeActiveCommunicationRequest(index);
```
Comment (achen2401) [inline src/components/ScheduleSetup.tsx:318]: @hannahburkhardt hmm.. looks like I am no longer able to remove a message. 
[see here](https://user-images.githubusercontent.com/12942714/224865537-fcd4172e-bbcc-4c42-935f-a2419824f419.webm)
I was able to in the [shared dev instance](https://femr.isacc.dev.cirg.uw.edu/home)

Comment (achen2401) [reply src/components/ScheduleSetup.tsx:318]: I _think_ the issue might be [here](https://github.com/uwcirg/isacc-messaging-client-sof/pull/12/files#diff-63ce71e720747f893a3baa1e98f3b9c83cdc625c9ae455f5aea8c1aac4669f95R127)

File: src/model/CarePlan.ts [hunk 3]
```diff
@@ -66,32 +66,69 @@ export default class CarePlan implements ICarePlan {
...
+    }
+
+    removeActiveCommunicationRequest(index: number) {
+        let requests = this.getActiveCommunicationRequests();
+        // if record doesn't exist yet on server, just delete it. Otherwise, don't delete, but rather, mark it as revoked
+        if (requests[index].id) {
+            requests[index].status = "revoked";
+            this.updateActivityAttr();
+        } else {
+            requests.splice(index, 1);
```
Comment (achen2401) [inline src/model/CarePlan.ts:127]: Hmm, this didn't seem to remove the record that doesn't exist yet.   I tried this `this.communicationRequests.splice(index, 1);` and that seems to work

File: src/components/EnrollmentApp.tsx (hunk 1 above)
Comment (hannahburkhardt) [reply src/components/EnrollmentApp.tsx:104]: Are you using R4? I think the value set was CarePlanStatus in R3, but both of the links you shared use RequestStatus for the status attribute in R4

Comment (achen2401) [reply src/components/EnrollmentApp.tsx:104]: ah, okay, I am using R4 and tested in R4.

File: src/model/CarePlan.ts (hunk 3 above)
Comment (hannahburkhardt) [reply src/model/CarePlan.ts:127]: Good catch! We do also need to update the activity attribute, so we need to call `this.setCommunicationRequests(requests)`

File: src/components/ScheduleSetup.tsx (hunk 2 above)
Comment (hannahburkhardt) [reply src/components/ScheduleSetup.tsx:318]: Thanks! fixed.

**Review by achen2401 — APPROVED** [review]

**Pull request merged**